### DISCUSSION TITLE
Using environment variables in vercel.json

### DISCUSSION THREAD

**Discussion by bar-lavie** [vercel/community#1734] [Help] [answered]
### Summary

Is there a way to pass Vercel env variable to vercel.json file?
vercel.json is part of my git repo and I believe it's a bad practice to expose it like that.
I want to use it to secure the corn job event:
```
  "crons": [
    {
      "path": "/api/backup-cron?backupType=daily&sharedKey=myUnsecureSharedKey", <--
      "schedule": "0 0 * * *"
    }
  ]
```

**Comment by wildky** [comment]
I am wondering the same thing. The [relevant docs](https://vercel.com/docs/cron-jobs#how-to-secure-cron-jobs) breeze over the fact that they are recommending dropping a secret key in a file that is meant to be managed in a code repo. 

I was super pumped to hear about vercel cron jobs, but I was _really_ hoping there would be some way to load environment variables into the cron job runner so we can use cron jobs securely.

**Comment by thompsonbear** [comment]
Agree completely and unfortunately it doesn't look like it's supported yet. Would be great if we could do something like this:
```
"crons": [{
    "path": "/api/cron",
    "schedule": "0 10 * * *",
    "searchParams": [{
        "key": "@my-secret-key"
    }]
}]
```

Then the request url would look like `/api/cron?key=mySecureSharedKey`

**Comment by vvo** [comment, marked as answer]
Hey there, protecting Cron Jobs on Vercel is now possible using this method:

1. Create an environment variable named `CRON_SECRET` in your Vercel Project
2. When present, we will invoke your Cron Job with an authorization header this way: `Authorization: Bearer $CRON_SECRET`
3. You can use this authorization header to verify the Cron Job was invoked by Vercel

Let me know if that's working well. We will announce this more broadly soon, along with updating docs.

**Reply by bar-lavie** [reply]
Hey @vvo, 
Thanks for the response and your solid solution! 
I've shipped it to CR and I'll update once it works on production, I've used Sentry to report each step along the way.

**Reply by bar-lavie** [reply]
@vvo, I'm sorry to say that I'm not getting the `Authorization` header in the request.
I've added the `CRON_SECRET` to my env. variables in Vercel dashboard as described

This is the part in my code that's validating the `CRON_SECRET`

```javascript
export function validateCronjobBearerToken(request: NextApiRequest) {
  const token = request.headers.authorization?.split(' ')[1];
  if (!token) {
    Sentry.captureEvent({
      level: 'debug',
      message: `Missing vercel Bearer token in request ${request.url}`,
      extra: {
        requestHeaders: request.headers,
      },
    });
    return false;
  }
  return token === CRON_JOBS.CRON_SECRET;
}
```

Hope I'm not missing anything

**Reply by AndyBitz** [reply]
@bar-lavie one question, is `CRON_JOBS.CRON_SECRET` correct in your example? Or should it be `process.env.CRON_SECRET`?

I also assume you redeployed after adding the environment variable?

Could you provide the Deployment URL for us to investigate further?
If not publicly, I'm happy to accept it via DMs on twitter https://twitter.com/andybitz_
Or via email `andy@vercel.com`.

**Reply by bar-lavie** [reply]
@AndyBitz `CRON_JOBS.CRON_SECRET` is just a function that exposes `process.env.CRON_SECRET` value behind the scenes

And yes, my code was redeployed after adding the env variable.

I've emailed you the deployment URL

**Reply by bar-lavie** [reply]
@AndyBitz the problem was me trying to trigger the corn job from the `Run` button on the Vercel dashboard - which does not submit this header (should be when this feature will be announced?)

I've been testing another cron job that was triggered programmatically and it works now as expected!

Thank you for your help! 😃

**Reply by wildky** [reply]
> @AndyBitz the problem was me trying to trigger the corn job from the `Run` button on the Vercel dashboard - which does not submit this header (should be when this feature will be announced?)
> 
> I've been testing another cron job that was triggered programmatically and it works now as expected!
> 
> Thank you for your help! 😃

I love a good corn job! 🌽 

(jokes aside, this is good to know about the dashboard testing, I would have likely done the same thing!)

**Reply by AndyBitz** [reply]
Thank you for your help with this one! Yes, it should work once announced, still something we need to add.

**Reply by AndyBitz** [reply]
👋 quick update: the `Run` button now supports this behaviour

**Comment by tianhuil** [comment]
@vvo: can we get this into the documentation?  I'd love for this to become a more stable documented feature.

**Comment by pm0u** [comment]
Going to assume this is finalized and just not announced? echo above, an official announcement/documentation would be great.

**Comment by pirtlj** [comment]
It's great to discover this,  the Auth story for cron jobs was a big missing piece.
